Support Policy for .NET Packages (#9206)
dotnet/core: package-support-policy.md

# Support Policy for .NET Packages

> .NET library packages provide additional functionality that complements the .NET libraries and application frameworks like ASP.NET Core. The library packages have the same [support model as the .NET versions](https://dotnet.microsoft.com/en-us/platform/support/policy) that they target.

Reviewer: ```suggestion
.NET library packages provide additional functionality that complements the .NET libraries and application frameworks like ASP.NET Core. The library packages have the same [support model as the .NET versions](https://dotnet.microsoft.com/platform/support/policy) that they target.
```

For example, [Microsoft.Extensions.Logging](https://www.nuget.org/packages/Microsoft.Extensions.Logging/#supportedframeworks-body-tab) supports multiple .NET versions via the frameworks it targets.

Our packages support any target framework versions that are currently in-support. You can find out which frameworks are supported by a specific package by visiting its NuGet page and navigating to the frameworks tab.

You can find out which target frameworks are currently in-support by visiting the following pages:

> - .NET: https://dotnet.microsoft.com/en-us/platform/support/policy

Reviewer: ```suggestion
- .NET: https://dotnet.microsoft.com/platform/support/policy
```

> - .NET Framework: https://dotnet.microsoft.com/en-us/learn/dotnet/what-is-dotnet-framework

Reviewer: ```suggestion
- .NET Framework: https://dotnet.microsoft.com/learn/dotnet/what-is-dotnet-framework
```

> - .NET Standard: https://learn.microsoft.com/en-us/dotnet/standard/net-standard

Reviewer: ```suggestion
- .NET Standard: https://learn.microsoft.com/dotnet/standard/net-standard
```

Keep in mind that the minimum .NET Standard version supported by a package depends on which minimum .NET and .NET Framework versions are supported by the package. Packages and projects targeting .NET Standard are expected to be compatible between each other by following the same rules specified by our .NET Standard policy.

## Release cadence

> Some packages follow the same release cadence as the .NET release cadence, meaning we release a new version of the package each time we release a new .NET version. For example:

Reviewer: ```suggestion
Some packages follow the same release cadence as the .NET release cadence, meaning a new version of the package is released each time a new .NET version is released. For example:
```

> - [System.Text.Json](https://www.nuget.org/packages/System.Text.Json), which is part of the shared framework, but is also available as a NuGet package.

Reviewer: ```suggestion
- [System.Text.Json](https://www.nuget.org/packages/System.Text.Json), which is part of the shared framework, but it's also available as a NuGet package.
```

Other packages have their own independent release cadence. These are packages that had their source code living in branches that are out of support, but the packages themselves are still in-support, because they target frameworks that are still supported. Some examples:

- [System.Xml.XPath.XmlDocument](https://www.nuget.org/packages/System.Xml.XPath.XmlDocument), which used to come out of the .NET Core 1.1 branch.
- [System.Buffers](https://www.nuget.org/packages/System.Buffers), which used to come out of the .NET Core 2.1 branch.
- [Microsoft.Bcl.HashCode](https://www.nuget.org/packages/Microsoft.Bcl.HashCode), which used to come out of the .NET Core 3.1 branch.
- [System.Net.WebSockets.WebSocketProtocol](https://www.nuget.org/packages/System.Net.WebSockets.WebSocketProtocol), which used to come out of the .NET 5.0 branch.

These, as well as many other packages, now have their source code living in https://github.com/dotnet/maintenance-packages. Any servicing releases required for these packages will come out of this repo.

## End of support

A package will be supported for as long as it targets frameworks that are still supported. For example:

> - [System.Numerics.Vectors](https://www.nuget.org/packages/System.Json#supportedframeworks-body-tab). Of all the included target frameworks in the package, .NET Standard 2.0 is still in support at the moment of writing this document. As long as there are frameworks compatible with .NET Standard 2.0, the package will remain in support.

Author: @ericstj is this an accurate representation of the time a package will be supported?

Reviewer 2: `NetStandard2.0` is not a runnable framework, so only having a `netstandard2.0` asset doesn't imply the package is in support.  The package would need an applicable implementation on a runnable, in support framework.

For `System.Numerics.Vectors` I think you determined it was in support because it had an implementation which was applicable for `net462`.  Since `net462` is still in support this package is still in support.

Author: > For System.Numerics.Vectors I think you determined it was in support because it had an implementation which was applicable for net462

Yes. When adding the `System.Numerics.Vectors` package to a project targeting `net462`, if I consume the APIs of that namespace, I can confirm that the DLL from the package is used (not the shared framework).

Okay, thanks for the explanation on applicability. I will rewrite that part.

## Reporting issues

Developers should report issues in the repo that owns the package, if they know which one it is (https://github.com/dotnet/runtime, https://github.com/dotnet/aspnetcore). When in doubt, we encourage you to report the issue directly in https://github.com/dotnet/runtime, and the repo maintainers will be happy to help transfer the issue to the rightful owners.

> There are some assemblies that used to be published out of a .NET version that is currently out of support, but the packages themselves are still supported because they still target a supported .NET version. For these few special cases, we host their source code in the https://github.com/dotnet/maintenance-packages repo. We still encourage users to report issues for these assemblies directly in the central repo https://github.com/dotnet/runtime.

Reviewer: ```suggestion
There are some assemblies that used to be published out of a .NET version that is currently out of support, but the packages themselves are still supported because they still target a supported .NET version. For these few special cases, their source code is hosted in the https://github.com/dotnet/maintenance-packages repo. We still encourage users to report issues for these assemblies directly in the central repo https://github.com/dotnet/runtime.
```

# Support Policy for .NET Packages

> NuGet packages owned by the .NET team are supported as long as they target at least one framework that is still actively supported.

Reviewer 3: ```suggestion
.NET library packages provide additional functionality that complements the .NET libraries and application frameworks like ASP.NET Core. The library packages have the same support model as the .NET versions that they target.

For example, [Microsoft.Extensions.Logging](https://www.nuget.org/packages/Microsoft.Extensions.Logging/#supportedframeworks-body-tab) supports multiple .NET versions via the frameworks it targets.
```

Reviewer 2: I'm not sure how to write it, but it might be helpful to distinguish which version of the package is supported in that way.  Not all versions - only *a* version.

We'll continue to support a version of the package, if that package was removed from the product release and it still targets in support frameworks (eg: System.Memory).  We'll also support an additional version of a package if future versions of that package removed targeting for an in support framework (eg: Asp.NET Core on netfx).

Author: ^ Still thinking of this feedback by Eric. I addressed all the other feedback.

Author: I see this doc has a "Servicing" and an "End of support" sections: https://github.com/dotnet/core/blob/main/release-policies.md#end-of-support

I'll try to integrate Eric's request based on that info.

> The minimum supported .NET versions are:

Reviewer 3: Is it necessary to include this information. We don't want more places to update.

Author: Do we have a page where we indicate what are the minimum supported .NET versions? I could link to that instead.

Reviewer 3: - https://github.com/dotnet/core/blob/main/releases.md
- https://dotnet.microsoft.com/platform/support/policy

- .NET Framework 4.6.2
- .NET Standard 2.0
- .NET 6.0

The support policy of our frameworks can be found here: https://dotnet.microsoft.com/en-us/platform/support/policy

> ## Package and project compatibility

Reviewer 3: This text is quite verbose. It occurs to me that the more terse text in our exist support doc almost says the same thing: https://github.com/dotnet/core/blob/main/release-policies.md#end-of-support

Author: I see. I can rewrite the section to just point to that document so users can consult the way packages are supported depending on the project and the package tfms.

Reviewer 3: I propose that we update that text to be a tad more expansive and then see if we need this document at all.

A package targeting any of these above mentioned supported frameworks is expected to be compatible with projects targeting any newer versions. For example:

> - If you have a project that targets .NET 8.0, and it consumes a package that targets either .NET Standard 2.0 or .NET 6.0, you should expect the package to be compatible with your project because .NET Standard 2.0 supports .NET 8.0 and .NET 8.0 is greater than .NET 6.0.

Reviewer 4: Maybe it could be made in table form intead of text?
like Package TFM, Project TFM, Comment

probably it will be easiear to understand

Author: Interesting suggestion. Although based on the rest of the feedback, it seems the version compatibility can be understood by looking at the .NET Standard document. I can just link to that and keep it simpler: no list and no table needed.

- If you have a project that targets .NET 7.0, and it consumes a package that targets .NET 6.0, you should expect the package to be compatible with the project because .NET 7.0 is greater than .NET 6.0.
- If you have a project that targets .NET Framework 4.8, and it consumes a package that targets .NET Standard 2.0 and .NET Framework 4.6.2, you should expect the package to be compatible with the project because .NET Standard 2.0 supports .NET Framework 4.8 and .NET Framework 4.8 is greater than .NET Framework 4.6.2.
- If you have a project that targets .NET 6.0 and it consumes a package that targets .NET 8.0, the package is not compatible with the project.
- If you have a project that targets .NET Framework 4.6.2 and it consumes a package that targets .NET Framework 4.8, the package is not compatible with the project.

The detailed list of .NET Frameworks supported by each .NET Standard version can be found here: https://learn.microsoft.com/en-us/dotnet/standard/net-standard

> ## Package types

Reviewer 3: A bunch of this content should be in docs. It isn't "support information".

Author: Where in docs would you suggest putting it? Or would this information be more useful if it was rewritten with a more precise/specific focus on support?

I thought this information was relevant to users from the perspective of the lifecycle of a package because it might be difficult to understand for cases where it's not clear to them what happened to (or where is) the latest version of the package.

I'll give you an example: Why is MS Docs showing new APIs for an assembly that I've been using as a NuGet package, but the latest version of said package does not include those APIs? Surprise surprise: turns out the assembly got absorbed by the shared framework and the latest version contains those new APIs.

Reviewer 3: I don't know where the library teams hosts general content in docs about packages. If there isn't such a place, now might be a good time to create one. Major point is that this information goes well beyond support.

> I'll give you an example: Why is MS Docs showing new APIs for an assembly that I've been using as a NuGet package, but the latest version of said package does not include those APIs? Surprise surprise: turns out the assembly got absorbed by the shared framework and the latest version contains those new APIs.

Sure. Let's separate out lifecycle (when a package goes EOL) from weird migration topics. They have a relationship, but are totally separate concerns. Support documents are intended to be super crisp and as terse as possible (while still readable/useful).

https://dotnet.microsoft.com/en-us/platform/support/policy/aspnet/2.1-packages

## Reporting issues

Developers should report issues in the repo that owns the package, if they know which one it is (https://github.com/dotnet/runtime, https://github.com/dotnet/aspnetcore). When in doubt, we encourage you to report the issue directly in https://github.com/dotnet/runtime, and the repo maintainers will be happy to help transfer the issue to the rightful owners.

There are some assemblies that used to be published out of a .NET version that is currently out of support, but the packages themselves are still supported because they still target a supported .NET version. For these few special cases, we host their source code in the https://github.com/dotnet/maintenance-packages repo. We still encourage users to report issues for these assemblies directly in the central repo https://github.com/dotnet/runtime.

The following table lists the special case assemblies that were migrated to maintenance-packages, their branch of origin, and the way they're supported in each framework category.

> | Assembly | Origin | 6.0+ | .NET Standard 2.0 | 4.6.2+ |

Reviewer 3: This is an interesting table for dotnet/maintenance-packages. It isn't clear to me what the support implications are.

Author: I see. What I want to achieve is letting users know what to expect in terms of support for packages that used to be built out of branches that are now out of support, but the packages themselves are still in support. I think the sentence in line 68 covers that, the table might not be necessary as you mention.

Reviewer 3: I think the table is useful, but I see it more as a novelty for <1% of users.
